src/ledger/LedgerManagerImpl.cpp

```diff
@@ -638,66 +641,15 @@ LedgerManagerImpl::historyCaughtup(asio::error_code const& ec,
-                // during catchup replay. We can just drop these, they're
-                // redundant with what catchup did.
-
-                if (lcd.getLedgerSeq() == mLastClosedLedger.header.ledgerSeq)
-                {
-                    // At the knit-up point between history-replay and
-                    // buffer-replay, we should have identity between the
-                    // contents of the consensus LCD and the last ledger catchup
-                    // closed (which was proposed as a candidate, and we
-                    // approved in verifyCatchupCandidate).
-                    assert(lcd.getTxSet()->getContentsHash() ==
-                           lcd.getValue().txSetHash);
-                    assert(lcd.getValue() == mLastClosedLedger.header.scpValue);
-                }
-
-                continue;
-            }
-            else if (lcd.getLedgerSeq() ==
-                     mLastClosedLedger.header.ledgerSeq + 1)
-            {
-                CLOG(INFO, "Ledger")
-                    << "Replaying buffered ledger-close: "
-                    << "[seq=" << lcd.getLedgerSeq() << ", prev="
-                    << hexAbbrev(lcd.getTxSet()->previousLedgerHash())
-                    << ", tx_count=" << lcd.getTxSet()->size()
-                    << ", sv: " << stellarValueToString(lcd.getValue()) << "]";
-                closeLedger(lcd);
-            }
-            else
-            {
-                // We should never _overshoot_ the last ledger. The whole point
-                // of rounding the initLedger value up to the next history
-                // boundary is that there should always be some overlap between
-                // the buffered LedgerCloseDatas and the history block we catch
-                // up with.
-                //
-                // So if we ever get here, something was seriously wrong --
-                // possibly SCP timed out / fell behind _during_ catchup -- and
-                // we should flush everything we did during catchup and restart
-                // the process anew.
-
-                assert(lcd.getLedgerSeq() >
-                       mLastClosedLedger.header.ledgerSeq + 1);
-                CLOG(ERROR, "Ledger")
-                    << "Catchup failed to buffer contiguous ledger chain";
-                CLOG(ERROR, "Ledger")
-                    << "LCL is " << ledgerAbbrev(mLastClosedLedger)
-                    << ", trying to apply buffered close " << lcd.getLedgerSeq()
-                    << " with txhash "
-                    << hexAbbrev(lcd.getTxSet()->getContentsHash());
-                mSyncingLedgers = {};
-                mSyncingLedgersSize.set_count(mSyncingLedgers.size());
-                CLOG(ERROR, "Ledger") << "Catchup will restart at next close.";
-                setState(LM_BOOTING_STATE);
-                return;
-            }
+            assert(lcd.getLedgerSeq() ==
```

Comment:
I don't see how this really works, there was a code path here that leads to `LM_BOOTING_STATE` - which cannot happen with the new code (as far as I can tell)?
Meaning that if we "catchup" to the checkpoint but there is a gap in history it will set the state to synced even though it's not (and what it logged "this round of catchup will fail and restart." will only happen after some timeout?).

Reply:
This is not possible because of how SyncingLedgerChain works - it can only contain contiguous chain of ledgers.

You can look at the version of LedgerManager before implementing SyncingLedgerChain - https://github.com/stellar/stellar-core/blob/9c29696964ad1268fe442bbf6ea5de17e3485d06/src/ledger/LedgerManagerImpl.cpp#L407

It was working the same way, so the only way it could happen even then was if the first ledger in mSyncingLedgers was overshooting. And it was impossible, because then verification would not pass at all. So this code was never used.

Instead when there is a gap, the catchup goes as far as it can, then it gets to SYNCED state until next ledger gets externalized and then it goes back to CATCHING_UP.

Reply:
I see. As catchup can take hours, there is a non trivial chance of having gaps: we should make it that at the end of catchup it waits for a ledger to close (probably by adding a new state `LM_FINISH_CATCHUP_STATE`), that way we can formalize the transition to "Synced!" and avoid triggering publish, etc when not fully in sync yet.